src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkTransition.java

```diff
@@ -329,11 +329,20 @@ private static Target getActual(
         if (allowsMultiple) {
           // if this setting allows multiple settings
           if (!(newValue instanceof List)) {
-            throw new TransitionException(
-                String.format(
-                    "'%s' allows multiple values and must be set"
-                        + " in transition using a starlark list instead of single value '%s'",
-                    actualSetting, newValue));
+            if (newValue.equals(rule.getAttr(STARLARK_BUILD_SETTING_DEFAULT_ATTR_NAME))) {
+              // We make an exception for the default value here for two reasons:
+              // 1. It ensures that specifying a setting with allow_multiple as an input but not as
+              //    an output of a transition does not run into the exception below.
+              // 2. It provides a clear way to reset the setting to its (literal) default from a
```

Comment:
2. here feels a bit API-shaky to me. Basically, a transition has to set a setting to a certain type unless it's set to a certain special value in which case it's exempt. I get the intended use case described here. But I can also see transitions that, say, read their values from attributes, which happen to set values to defaults - or not - by coincidence. i.e. whoever's setting the values has no insight into this and may be confused why some arbitrary string works while another fails.

To the point, do you think https://github.com/bazelbuild/bazel/pull/14911 obsoletes all this complexity?

Reply:
I agree it looks very hacky and is another reason to consider #14972 the better kind of fix. That said, I don't think this accidental resetting would be very realistic: The "magic" literal default value is the only valid one of its type (string rather than list of string) due to type validation and any attribute would only ever yield one type.

> To the point, do you think https://github.com/bazelbuild/bazel/pull/14911 obsoletes all this complexity?

The new representation of repeated string flags would no longer require any special handling for transitions, so we could eventually delete the entire `allowsMultiple` branch. But given that finalizing the design of #14911 will take some more time, either this PR or #14972 are probably needed to deal with the regression for 5.1.0.

Reply:
Agreed on your last point - I'm happy to fast track a temporary fix for the regression. I just wanted to verify the long-term expected state.